lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+                raise Exception(e)
+
+            # Put artifact in the right session group
+            artifacts_by_session[art._session_id] = artifacts_by_session.get(
+                art._session_id, []
+            ) + [art]
+
+        # For each session, construct SessionArtifacts object
+        for session_id, session_artifacts in artifacts_by_session.items():
+            self.session_artifacts[session_id] = SessionArtifacts(
+                session_artifacts
+            )
+
+    def _sort_session_artifacts(
+        self, dependencies: TaskGraphEdge = {}
+    ) -> List[SessionArtifacts]:
+        """
+        Use the user-provided artifact dependencies to
+        topologically sort a list of SessionArtifacts objects.
+        Raise an exception if the graph contains a cycle.
+        """
+        # Construct a combined graph across multiple sessions
+        combined_graph = nx.DiGraph()
+        for session_artifacts in self.session_artifacts.values():
+            session_graph = session_artifacts.graph.nx_graph
+            combined_graph.add_nodes_from(session_graph.nodes)
+            combined_graph.add_edges_from(session_graph.edges)
+
+        # Augment the graph with user-specified edges
+        dependency_edges = list(
+            chain.from_iterable(
+                (
+                    (
+                        self.art_name_to_node_id.get(artname, None),
+                        self.art_name_to_node_id.get(to_artname, None),
+                    )
+                    for artname in from_artname
+                )
+                for to_artname, from_artname in dependencies.items()
+            )
+        )
+        if None in list(chain.from_iterable(dependency_edges)):
+            raise KeyError(
+                "Dependency graph includes artifacts not in this artifact collection."
+            )
+        combined_graph.add_edges_from(dependency_edges)
+
+        # Check if the graph is acyclic
+        if nx.is_directed_acyclic_graph(combined_graph) is False:
+            raise Exception("Provided dependencies result in a cyclic graph.")
+
+        # Identify topological ordering between sessions
+        session_id_nodes = list(self.session_artifacts.keys())
+        session_id_edges = []
+        for node_id, to_node_id in dependency_edges:
+            assert node_id is not None
+            assert to_node_id is not None
+            from_session_id = self.node_id_to_session_id.get(node_id, None)
+            to_session_id = self.node_id_to_session_id.get(to_node_id, None)
+            if from_session_id is not None and to_session_id is not None:
+                session_id_edges.append((from_session_id, to_session_id))
+        inter_session_graph = nx.DiGraph()
+        inter_session_graph.add_nodes_from(session_id_nodes)
+        inter_session_graph.add_edges_from(session_id_edges)
+        session_id_sorted = nx.topological_sort(inter_session_graph)
```

Comment:
I don't think this will always work. The DAG check above is for the graph where each node is a LineaNode ... I agree that graph should be a DAG and topologic sortable. 

InterSessionGraph is a graph where each node is a session right? What if the user specifies dependencies like 
Artifact A (Session 1) -> Artifact B (Session 2)
Artifact B (Session 2) -> Artifact C (Session 1)
This would cause session 1 to depend on Session 2 and vice versa in inter_session_graph which is a cycle you cant topo sort. 

In fact this brings up a bigger issue which is that I don't think we can topologically sort the session IDs at all because a valid execution order might necessarily jump between sessions. In the example above, we would get the artifacts in order A,B,C but at a session level that requires us to jump from session 1, to session 2, and come back to session 1.

Reply:
@andycui97 Yes, you are right that `InterSessionGraph` is a graph of session IDs; hence, what the code is checking is if the session IDs form a DAG.

@mingjerli and I discussed the type of back-and-forth "hopping" between sessions that you described, but we concluded that this type of dependency should not occur in practice because each session is sort of a self-contained development environment. That is, given that development happens from session to session (i.e., development is very unlikely to happen across multiple simultaneous sessions), the sessions should form a DAG and be topologically sortable.

Hence, we are thinking/assuming that the example scenario you described (copied below) would not arise in practice; for this type of dependency to make sense, Artifacts A, B, and C should have had this developmental sequence in the first place, in which case the circular interspersion between sessions would not have happened.
```
Artifact A (Session 1)  >>  Artifact B (Session 2)  >>  Artifact C (Session 1)
```

Reply:
@andycui97 I am rethinking this. Could you share a concrete example for the scenario you described, i.e.,
```
Artifact A (Session 1) -> Artifact B (Session 2)
Artifact B (Session 2) -> Artifact C (Session 1)
```

Reply:
This is not an ideal use case but it might create the above: 

User has two notebooks, one of them is the main notebook (session 1) that prepares the training dataset Artifact A. 
Once this training set is created, they switch to a separate notebook (session 2) that pulls the standard test set Artifact B from their organization and writes it locally. Then they switch back to the main notebook , load the local test set, and call model.fit_predict(train_set, test_set) and model becomes Artifact C. 

The user may think that the order of dependencies to generate these artifacts needs to be training_data >> test_data >> model

Although I would argue that yes technically in this case you can re-order to test_data >> training_data >> model, but would the user think that way?

Reply:
@andycui97 I think we need a mechanism to merge graphs between multiple sessions to achieve what you describe but I don't think it is in the current scope. I actually have some idea about how to achieve that, I will start a design doc and everyone in the team can have a deeper discussion. In the current implementation, I think we can only call it out loud that there is a limitation to working with multiple sessions.

Reply:
Had a chance to sync w/ @mingjerli again about topological sorting between sessions. We seem to still prefer imposing topological ordering because the current implementation of graph refactor (into non-overlapping functions) demands we order sessions in a linear fashion (i.e., we cannot yet merge/intersperse different session graphs). As mentioned above, we need to implement inter-session graph merge if we want to support circular dependencies like the one we discussed (`Artifact A (Session 1) -> Artifact B (Session 2) -> Artifact C (Session 1)`). So, for now, we can proceed with the current solution while noting/describing this issue clearly (in both the user-facing error message and the code documentation).

Reply:
sgtm